python/ray/train/examples/jax_mnist_example.py

```diff
@@ -0,0 +1,220 @@
+    # train_ds["image"] = train_ds["image"][
+    #     : len(train_ds["image"]) // jax.device_count() * jax.device_count()
+    # ]
+    # train_ds["label"] = train_ds["label"][
+    #     : len(train_ds["label"]) // jax.device_count() * jax.device_count()
+    # ]
+    # test_ds["image"] = test_ds["image"][
+    #     : len(test_ds["image"]) // jax.device_count() * jax.device_count()
+    # ]
+    # test_ds["label"] = test_ds["label"][
+    #     : len(test_ds["label"]) // jax.device_count() * jax.device_count()
+    # ]
+
+    train_ds["image"] = np.float32(shard_fn(train_ds["image"])) / 255.0
+    test_ds["image"] = np.float32(shard_fn(test_ds["image"])) / 255.0
+    train_ds["label"] = np.int32(shard_fn(train_ds["label"]))
+    test_ds["label"] = np.int32(shard_fn(test_ds["label"]))
+    return train_ds, test_ds
+
+
+def train_func(config: Dict):
+    import jax
+    import jax.numpy as jnp
+    from flax import jax_utils
+    from flax import linen as nn
+    from flax.training import train_state
+    from flax.training.common_utils import shard
+    from jax import lax
+
+    # NOTE: the flax nn module has to define inside
+    # otherwise, the error message `ValueError: parent must be None, Module or Scope`
+    # see: https://github.com/google/flax/discussions/1390
+    class MLP(nn.Module):
+        """A simple mlp model."""
+
+        @nn.compact
+        def __call__(self, x):
+            x = x.reshape((x.shape[0], -1))  # flatten
+            x = nn.Dense(features=512)(x)
+            x = nn.relu(x)
+            x = nn.Dense(features=256)(x)
+            x = nn.relu(x)
+            x = nn.Dense(features=10)(x)
+            return x
+
+    @functools.partial(jax.pmap, static_broadcasted_argnums=(1, 2))
+    def create_train_state(rng, learning_rate, momentum):
+        """Creates initial `TrainState`."""
+        mlp = MLP()
+        params = mlp.init(rng, jnp.ones([1, 28, 28, 1]))["params"]
+        tx = optax.sgd(learning_rate, momentum)
+        return train_state.TrainState.create(apply_fn=MLP().apply, params=params, tx=tx)
+
+    @functools.partial(jax.pmap, axis_name="ensemble")
+    def train_step(state, images, labels):
+        """Computes gradients, loss and accuracy for a single batch."""
+
+        def loss_fn(params):
+            logits = MLP().apply({"params": params}, images)
+            one_hot = jax.nn.one_hot(labels, 10)
+            loss = jnp.mean(optax.softmax_cross_entropy(logits=logits, labels=one_hot))
+            return loss, logits
+
+        grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+        (loss, logits), grads = grad_fn(state.params)
+        grads = lax.pmean(grads, axis_name="ensemble")
+        accuracy = jnp.mean(jnp.argmax(logits, -1) == labels)
+        state = state.apply_gradients(grads=grads)
+        loss = lax.pmean(loss, axis_name="ensemble")
+        accuracy = lax.pmean(accuracy, axis_name="ensemble")
+        return state, loss, accuracy
```

Comment:
Logic looks right to me, but can you add a reference to how you got to this? E.g. if there is a flax example.

Reply:
sgtm